core/include/drivers/clk.h

```diff
@@ -58,32 +85,62 @@ struct clk_ops {
 			       unsigned long parent_rate);
 	unsigned long (*get_rate)(struct clk *clk,
 				  unsigned long parent_rate);
+	const char *(*get_name)(struct clk *clk);
 };
 
+/*
+ * Helper to identify clock operator type
+ */
+static inline bool clk_is_clk_elt(struct clk *clk)
+{
+	return clk->ops->id == CLK_OPS_ELT;
+}
+
+static inline struct clk_elt *clk_to_clk_elt(struct clk *clk)
+{
+	assert(clk_is_clk_elt(clk));
+
+	return (struct clk_elt *)clk;
+}
+
+static inline bool clk_is_clk_orphan(struct clk *clk)
+{
+	return clk->ops->id == CLK_OPS_ORPHAN;
+}
+
 /**
  * Return the clock name
  *
  * @clk: Clock for which the name is needed
  * Return a const char* pointing to the clock name
  */
-static inline const char *clk_get_name(struct clk *clk)
-{
-	return clk->name;
-}
+const char *clk_get_name(struct clk *clk);
 
 /**
- * clk_alloc - Allocate a clock structure
+ * clk_alloc - Allocate a clock element structure
  *
  * @name: Clock name
  * @ops: Clock operations
  * @parent_clks: Parents of the clock
  * @parent_count: Number of parents of the clock
  *
- * Returns a clock struct properly initialized or NULL if allocation failed
+ * Returns a struct clk * or NULL if allocation failed.
+ * The return address actually points to a struct clk_elt instance.
+ * One can use clk_to_clk_elt() to convert the reference type.
  */
 struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
 		      struct clk **parent_clks, size_t parent_count);
 
+/**
+ * clk_alloc_orphans - Allocate an array of orphan clocks (1 or more)
+ *
+ * @ops: Clock operations
+ * @count: Number of clocks
+ *
+ * Returns a clock struct properly initialized or NULL if allocation failed
+ */
+struct clk *clk_alloc_orphans(const struct clk_ops *ops, size_t count);
```

Comment:
Can't `clk_alloc` be used with  num_parent = 0 ? This would allow to keep a single function accross different drivers.